avail-server/src/main/kotlin/com/avail/server/io/files/ServerFileWrapper.kt

```diff
@@ -0,0 +1,409 @@
+
+	/**
+	 * The [queue][ConcurrentLinkedQueue] of [FileRequestHandler]s that are
+	 * requesting receipt of the [file]'s
+	 * [raw bytes][AvailServerFile.rawContent].
+	 */
+	private val handlerQueue = ConcurrentLinkedQueue<FileRequestHandler>()
+
+	/**
+	 * The [Stack] of [TracedAction] that tracks the [FileAction]s applied in
+	 * Stack order. To revert, pop the `TracedAction`s and apply
+	 * [TracedAction.reverseAction] to the [AvailServerFile]. This represents
+	 * the _undo_ stack.
+	 */
+	private val tracedActionStack = Stack<TracedAction>()
+
+	/**
+	 * The next position in the [tracedActionStack] to begin saving to disk the
+	 * next time the [file]'s local history is saved.
+	 */
+	private var tracedActionStackSavePointer: Int = 0
+
+	/**
+	 * Save the [tracedActionStack] to local history starting from the position
+	 * at [tracedActionStackSavePointer].
+	 */
+	fun conditionallySaveToLocalHistory ()
+	{
+		val currentSize = tracedActionStack.size
+		if (tracedActionStackSavePointer < currentSize)
+		{
+			val start = tracedActionStackSavePointer
+			for (i in start until  currentSize)
+			{
+				// TODO write this somewhere!
+				tracedActionStack[i]
+			}
+			tracedActionStackSavePointer = currentSize
+		}
+	}
+
+	/**
+	 * The [queue][ConcurrentLinkedQueue] of [FileAction]s that are waiting to
+	 * be executed for the contained [AvailServerFile].
+	 */
+	private val actionQueue = ConcurrentLinkedQueue<FileAction>()
+
+	/**
+	 * Is the [file] being available for [FileAction]a from the [actionQueue]?
+	 * `true` indicates it is; `false` otherwise.
+	 */
+	private val isReady = AtomicBoolean(true)
+
+	/**
+	 * Remove all the [FileAction]s from the [actionQueue] and perform them
+	 * in FIFO order if this [ServerFileWrapper] [is ready][isReady]
+	 * to update.
+	 */
+	private fun executeActions ()
+	{
+		if (isReady.getAndSet(false))
+		{
+			var action = actionQueue.poll()
+
+			// Check if an action is available and that the file hasn't been
+			// marked as being not ready before proceeding
+			while (action != null && isReady.get())
+			{
+				val tracedAction =
+					action.execute(file, System.currentTimeMillis())
+
+				if (undoStackDepth > 0)
+				{
+					// Must remove all TracedActions above the undoStackDepth
+					// pointer as they are no longer valid.
+					do
+					{
+						val ta = tracedActionStack.pop()
+						// TODO remove ta from the project cache?
+					}
+					while (--undoStackDepth > 0)
+				}
+				tracedActionStack.push(tracedAction)
+				action = actionQueue.poll()
+			}
+			// TODO is this safe
+			synchronized(actionQueue)
+			{
+				isReady.set(true)
+			}
+			// TODO Save file
+		}
+	}
+
+	/**
+	 * [Update][FileAction.execute] the wrapped [AvailServerFile] with the
+	 * provided [FileAction].
+	 *
+	 * @param fileAction
+	 *   The `FileAction` to perform.
+	 */
+	fun execute (fileAction: FileAction)
+	{
+		synchronized(actionQueue)
+		{
+			actionQueue.add(fileAction)
+		}
+		executeActions()
+	}
```

Comment:
The concurrency logic is messy in `ServerFileWrapper`, and I think we need to sit down and talk it through. It's probably not a lot of work to fix up, but we need to have consensus on strategic goals first. My gut feeling is that `isReady` needs to go completely, and needing to synchronized on a `ConcurrentLinkedQueue` is pretty fishy. It's also pretty suspicious that you queue an action and then immediately execute all actions, draining the queue.